src/core/search/ast_expr.h

```diff
@@ -4,21 +4,106 @@
 
 #pragma once
 
+#include <algorithm>
+#include <iostream>
+#include <memory>
 #include <ostream>
+#include <vector>
+
+#include "base/logging.h"
 
 namespace dfly {
 
 namespace search {
 
-class AstExpr {};
+// Describes a single node of the filter AST tree.
+class AstNode {
+ public:
+  virtual ~AstNode() = default;
+
+  // Check if this input is matched by the node.
+  virtual bool Check(std::string_view input) const = 0;
+
+  // Debug print node.
+  virtual std::string Debug() const = 0;
+};
+
+using AstNodePtr = std::unique_ptr<AstNode>;
+
+// Do we have to pollute all our parser file with moves?
+struct AstExpr {
+  AstExpr() : ptr{} {
```

Comment:
I'd just do `AstExpr() = default` (no need to specially initialize `ptr` as it's a `unique_ptr` which points to `nullptr` by default)
Same goes for copy c'tor and move c'tor below